torchtune/utils/memory.py

```diff
@@ -15,14 +16,67 @@
 from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
     apply_activation_checkpointing,
 )
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 from torch.distributed.fsdp.wrap import ModuleWrapPolicy
+from torchtune import modules
 from torchtune.utils.logging import get_logger
 
 _log: logging.Logger = get_logger()
 
 ACWrapPolicyType: Type = Union[Set[Type], Callable[[nn.Module, bool, int], bool]]
 
 
+def get_ac_policy(model_type: str, modules_to_wrap: Set[Type]) -> ACWrapPolicyType:
+    """
+    Get the activation checkpointing policy for the model.
+
+    Args:
+        model_type (str): Model type.
+        modules_to_wrap (Set[Type]): Set of module types to wrap.
+
+    Returns:
+        ACWrapPolicyType: Policy to wrap module.
+    """
+    if model_type == "LLAMA3":
+        # ensure torch.nn.Embedding activations are checkpointed.
+        modules_to_wrap.add(torch.nn.Embedding)
+        return partial(_llama3_ac_policy, modules_to_wrap=modules_to_wrap)
+    else:
+        return ModuleWrapPolicy(modules_to_wrap)
+
+
+def _maybe_fsdp_unwrap(module: nn.Module) -> nn.Module:
+    """
+    Unwrap a module if it is wrapped by FSDP.
+
+    Args:
+        module (nn.Module): Module to unwrap.
+
+    Returns:
+        nn.Module: Unwrapped module.
+    """
+    if isinstance(module, FSDP):
+        return module.module
+    return module
+
+
+def _llama3_ac_policy(module: nn.Module, recurse: bool, modules_to_wrap, **kwargs):
+    # Label that output_proj should be wrapped individually.
+    if isinstance(module, modules.TransformerDecoder):
+        targ = _maybe_fsdp_unwrap(module.output)
```

Comment:
So this isn't really because we want to support FSDP and non FSDP models, but we do happen to have this support with this function.

The reason we need this unwrap call is if the model, when we wrap in AC, has already been wrapped in FSDP. Then we want to AC wrap `module.output`, but just accessing `module.output` when the model is FSDP wrapped may give us back the FSDP class. So we further unwrap to retrieve the local nn.module.